dali/python/nvidia/dali/external_source.py

```diff
@@ -62,31 +63,56 @@ def __next__(self):
             return next(self.it)
 
 class _ExternalSourceGroup(object):
-    def __init__(self, callback, is_multioutput, instances = [], cuda_stream = None, use_copy_kernel = None):
+    def __init__(self, callback, is_multioutput, instances = [], cuda_stream = None, use_copy_kernel = None, batch = True):
         self.instances = list(instances)  # we need a copy!
         self.is_multioutput = is_multioutput
         self.callback = callback
         self._cuda_stream = cuda_stream
         self.use_copy_kernel = use_copy_kernel
+        self.batch = batch
+        self.current_iter = 0
+        self.current_sample = 0
         if callback is not None:
             if callback.__code__.co_argcount not in [0, 1]:
                 raise TypeError("External source callback must be a callable with 0 or 1 argument")
-            self.accepts_iter_num = (callback.__code__.co_argcount == 1)
-        else:
-            self.accepts_iter_num = None
+            self.accepts_arg = callback.__code__.co_argcount > 0
 
     def append(self, instance):
         self.instances.append(instance)
 
-    def call_and_feed(self, pipeline, current_iter):
-        if self.accepts_iter_num:
-            callback_out = self.callback(current_iter)
+    def callback_args(self, idx_in_batch):
+        if not self.accepts_arg:
+            return ()
+        if idx_in_batch is not None:
+            arg = nvidia.dali.types.SampleInfo(self.current_sample + idx_in_batch, idx_in_batch, self.current_iter)
         else:
-            callback_out = self.callback()
+            arg = self.current_iter
+        return (arg,)
+
+    def reset_indices(self):
+        self.current_iter = 0
+        self.current_sample = 0
+
+    def call_and_feed(self, pipeline, batch_size):
+        try:
+            if self.batch:
+                callback_out = self.callback(*self.callback_args(None))
+            else:
+                callback_out = [self.callback(*self.callback_args(i)) for i in range(batch_size)]
+            self.current_sample += batch_size
+            self.current_iter += 1
+        except StopIteration:
+            self.current_sample = 0
```

Comment:
Could it be self.reset_indices() call?

Reply:
It could. I guess I won't re-run CI just for that, so I'll wait for another review - if there are some more changes requested, I'll adjust this one.